actor/todo/src/lib.rs

```diff
@@ -0,0 +1,282 @@
+use wasmcloud_interface_keyvalue::{
+    IncrementRequest, KeyValue, KeyValueSender, SetAddRequest, SetDelRequest, SetRequest,
+};
+#[derive(Debug, Default, Actor, HealthResponder)]
+#[services(Actor, HttpServer)]
+struct TodoActor {}
+
+// FIXME: contribute this upstream
+struct HttpResponseShim {}
+impl HttpResponseShim {
+    /// Creates a response with a given status code and serializes the given payload as JSON
+    pub fn json<T>(payload: T, status_code: u16) -> HttpResponse
+    where
+        T: Serialize,
+    {
+        HttpResponse {
+            body: serde_json::to_string(&payload).unwrap().into_bytes(),
+            header: HashMap::new(),
+            status_code,
+        }
+    }
+    /// Handy shortcut for creating a 404/Not Found response
+    pub fn not_found() -> HttpResponse {
+        HttpResponse {
+            status_code: 404,
+            ..Default::default()
+        }
+    }
+    /// Useful shortcut for creating a 200/OK response
+    pub fn ok() -> HttpResponse {
+        HttpResponse {
+            status_code: 200,
+            ..Default::default()
+        }
+    }
+
+    /// Useful shortcut for creating a 500/Internal Server Error response
+    pub fn internal_server_error(msg: &str) -> HttpResponse {
+        HttpResponse {
+            status_code: 500,
+            body: msg.to_string().as_bytes().into(),
+            ..Default::default()
+        }
+    }
+    // and also:
+    // /// Shortcut for creating a 400/Bad Request response
+    // pub fn bad_request() -> HttpResponse {
+    //     HttpResponse {
+    //         status_code: 400,
+    //         ..Default::default()
+    //     }
+    // }
+}
+
+/// Implementation of HttpServer trait methods
+#[async_trait]
+impl HttpServer for TodoActor {
+    async fn handle_request(&self, ctx: &Context, req: &HttpRequest) -> RpcResult<HttpResponse> {
+        let trimmed_path = req.path.trim_end_matches('/');
+        trace!("incoming req: {:?}, path: {:?}", req, trimmed_path);
+
+        match (req.method.as_ref(), trimmed_path) {
+            ("GET", "/") => Ok(HttpResponse {
+                body: "todo server lives at /api".to_string().into_bytes(),
+                ..Default::default()
+            }),
+
+            ("POST", "/api") => {
+                create_todo(
+                    ctx,
+                    serde_json::from_slice(&req.body).map_err(
+                        // FIXME: this is very sad.
+                        |e| e.to_string(),
+                    )?,
+                )
+                .await
+                .map(|todo| HttpResponseShim::json(todo, 200))
+            }
+
+            ("GET", "/api") => get_all_todos(ctx)
+                .await
+                .map(|todos| HttpResponseShim::json(todos, 200)),
+
+            ("GET", url) => get_todo(ctx, url)
+                .await
+                .map(|todo| HttpResponseShim::json(todo, 200))
+                .or_else(|_| Ok(HttpResponseShim::not_found())),
```

Comment:
TODO: it might be possible to let the keyvalue store errors bubble up one level, and match on them in a central place